pkg/querier/queryrange/series_volume.go

```diff
@@ -0,0 +1,216 @@
+				len(jobs),
+				len(jobs),
+				func(ctx context.Context, i int) error {
+					bucket, resp, err := jobs[i](ctx)
+					collector.Add(bucket, resp.(*VolumeResponse))
+					return err
+				})
+
+			if err != nil {
+				return nil, err
+			}
+
+			promResp := toPrometheusResponse(collector.responses)
+			return promResp, nil
+		})
+	})
+}
+
+type concurrentVolumeResponseCollector struct {
+	lock      sync.Mutex
+	responses map[time.Time]*VolumeResponse
+}
+
+func newConcurrentVolumeResponseCollector(capacity int) *concurrentVolumeResponseCollector {
+	return &concurrentVolumeResponseCollector{
+		lock:      sync.Mutex{},
+		responses: make(map[time.Time]*VolumeResponse, capacity),
+	}
+}
+
+func (c *concurrentVolumeResponseCollector) Add(bucket time.Time, resp *VolumeResponse) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.responses[bucket] = resp
+}
+
+func toPrometheusResponse(resps map[time.Time]*VolumeResponse) *LokiPromResponse {
+	var headers []*definitions.PrometheusResponseHeader
+	samplesByName := make(map[string][]logproto.LegacySample)
+
+	for bucket, resp := range resps {
+		if headers == nil {
+			headers := make([]*definitions.PrometheusResponseHeader, len(resp.Headers))
+			for i, header := range resp.Headers {
+				h := header
+				headers[i] = &h
+			}
+		}
+
+		for _, volume := range resp.Response.Volumes {
+			if _, ok := samplesByName[volume.Name]; !ok {
+				samplesByName[volume.Name] = make([]logproto.LegacySample, 0, 1)
+			}
+
+			samplesByName[volume.Name] = append(samplesByName[volume.Name], toPrometheusSample(volume, bucket))
+		}
+	}
+
+	promResponse := queryrangebase.PrometheusResponse{
+		Status:  loghttp.QueryStatusSuccess,
+		Data:    toPrometheusData(samplesByName),
+		Headers: headers,
+	}
+
+	return &LokiPromResponse{
+		Response:   &promResponse,
+		Statistics: stats.Result{},
+	}
+}
+
+func toPrometheusSample(volume logproto.Volume, t time.Time) logproto.LegacySample {
+	ts := model.TimeFromUnix(t.Unix())
+	return logproto.LegacySample{
+		Value:       float64(volume.Volume),
+		TimestampMs: ts.UnixNano() / 1e6,
+	}
+}
+
+type sortableSampleStream struct {
+	labels  labels.Labels
+	samples []logproto.LegacySample
+}
+
+func toPrometheusData(series map[string][]logproto.LegacySample) queryrangebase.PrometheusData {
+	resultType := loghttp.ResultTypeVector
+	sortableResult := make([]sortableSampleStream, 0, len(series))
+
+	for name, samples := range series {
+		if resultType == loghttp.ResultTypeVector && len(samples) > 1 {
+			resultType = loghttp.ResultTypeMatrix
+		}
+
+		lbls, err := syntax.ParseLabels(name)
+		if err != nil {
+			continue
+		}
+
+		sort.Slice(samples, func(i, j int) bool {
+			return samples[i].TimestampMs < samples[j].TimestampMs
+		})
+
+		sortableResult = append(sortableResult, sortableSampleStream{
+			labels:  lbls,
+			samples: samples,
+		})
+	}
+
+	sort.Slice(sortableResult, func(i, j int) bool {
+		return sortableResult[i].labels.String() < sortableResult[j].labels.String()
```

Comment:
It might be good to store the label string on the sortable type too. `.String()` allocates so this is surprisingly expensive if there are a lot of results.